cql3/statements/describe_statement.cc

```diff
@@ -142,6 +143,82 @@ future<std::vector<description>> generate_descriptions(
     co_return descs;
 }
 
+template <std::ranges::range Range>
+    requires std::is_base_of_v<keyspace_element, std::ranges::range_value_t<Range>>
+future<std::vector<description>> generate_descriptions(replica::database& db, const Range& elements,
+    std::optional<bool> with_internals = std::nullopt, bool sort_by_name = true)
+{
+    std::vector<description> descs{};
+
+    // In general, a range is allowed to be modified during the call to `std::ranges::size()`.
+    // However, we rely on the sanity of the passed types and the fact that we take the range
+    // as a const-reference.
```

Comment:
It's not necessarily that unlikely. An example could be a wrapper for a file descriptor that allows for iterating over consecutive tokens in the file. `std::ranges::size()` would basically iterate over the whole range, but since the range consumes tokens, we would be left with nothing. I'm not sure how people might see the `const` qualifier in cases like that (i.e. I can imagine that e.g. an object representing a file descriptor might not necessarily need to modify itself at all) and that's why I left that comment.

Reply:
If you pass a range that represents a file then I would assume that you made sure that nobody will modify the file in the meantime. Otherwise, you could have two unsychronized processes, one reading and another writing to the file, which sounds like a recipe for something that resembles a data race.

This function is clearly meant to read from the range, so passing such a misbehaving range to the function sounds like a bug on the caller side. I really think is is reasonable to assume that the range will be immutable.

Reply:
It's not necessarily about the file being modified while we're reading from it; more about the iterator corresponding to the range being single-use. Calling `std::ranges::size` consumes the iterator and it'll never be usable again. But I get your point; perhaps I'm trying to play it too safe here and it's okay to assume that the range types we get in the function behave correctly. I could even go a bit further and make it accept `std::vector` only, which would solve the issue altogether.